### PR TITLE
First draft of New Connection Documentation

### PR DESCRIPTION
**Pull request by shannonlal**
This is a first draft of the new Connection document as discussed in issue #266.  I have tried to break down all the major steps that are required to create a new connection.  I would welcome some comments and insight on how to make this document better.  In addition, I have put some instructions on how to run the development environment as TBD.  If someone can add some updated instructions that would be great.

### CODE REVIEW

File: NEW_CONNECTION.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Add New Data Source
+
+The following are the instructions for adding a new data source 
+
+## Add the following files for Creating a new Datasource
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: I'd format this as a list and perhaps have two sections: one for new files and another for updated files.

File: NEW_CONNECTION.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+backend/persistent/datastores/NameOfDatastoreToConnect.js
+app/constants/constants.js
+app/images
+app/actions/sessions.js
+app/components/Settings/Tabs/Tab.react.js
+app/components/Settings/Settings.react.js
+
+
+## Images
+It is recommended to add the images as png and a good idea to include a regular and small icon
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: Add locations were logos are currently stored.

File: NEW_CONNECTION.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+ * two elements.  The first is an array of strings which is the column names
+ * The second is a two dimensional array which represents the rows to be displayed
+ * @param {string|object} queryObject
+ * @param {object} connection
+ * @return {object} -- tuple {columnames, rows}
+ */ 
+function query(queryObject, connection); 
+
+//Question are these required?
+function files(connection);
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: remove (normal connectors don't use this)

File: NEW_CONNECTION.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+function storage(connection); 
+```
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: remove (normal connectors don't use this)

**Comment by n-riesco** [comment]
I'll give you more feedback after Jan, 3rd

**Comment by shannonlal** [comment]
I cleaned up some instructions to broke it down into lists as suggested.  A couple of points that I think would help myself and other developers.  
1. Instructions on how to setup the development and debugging environment.  
2. I am still working my way through the Athena connector and at this point I have not had to add anything to the Settings.react.js file.  I am not sure if I have missed anything at this point but any advice would be greatly appreciated.

**Comment by shannonlal** [comment]
Just a small update for the Settings.react.js file.  I figured out there was an update that needed to be added to this file to make the query work correctly

**Comment by chriddyp** [comment]
Could we update `CONTRIBUTING.md` with a link to this file?

File: NEW_CONNECTION.md [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+2. app/actions/sessions.js
+3. app/components/Settings/Tabs/Tab.react.js
+4. app/components/Settings/Settings.react.js
+5. backend/persistent/datastores/Datastores.js
+
+## The following are the new files that need to be added
+
+1. app/images
+2. backend/persistent/datastores/NameOfDatastoreToConnect.js
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: Most of the project uses lowercase files (and I intend to get rid of the expections):
```md
1. app/images/my-logo.png
2. backend/persistent/datastores/my-connector.js
```

File: NEW_CONNECTION.md [hunk 6]
```diff
@@ -0,0 +1,102 @@
+# Add New Data Source
+
+The following are the instructions for adding a new data source 
+
+## The following are the files that need to be updated
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: Now that you've provided detailed instructions, this section could be removed.

File: NEW_CONNECTION.md [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+1. app/constants/constants.js
+2. app/actions/sessions.js
+3. app/components/Settings/Tabs/Tab.react.js
+4. app/components/Settings/Settings.react.js
+5. backend/persistent/datastores/Datastores.js
+
+## The following are the new files that need to be added
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: Now that you've provided detailed instructions, this section could be removed.

File: NEW_CONNECTION.md [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+1. app/images
+2. backend/persistent/datastores/NameOfDatastoreToConnect.js
+
+# Detailed Implementation Instructions
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: I'd remove this section and have all the subsections under `# Add New Data Source`.

File: NEW_CONNECTION.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+## Images
+It is recommended to add the images as png and a good idea to include a regular and small icon.
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: We don't use the small icons any longer.
We only need an icon that looks good when displayed with `max-width: 100px;` and `max-height: 70px;`.

File: NEW_CONNECTION.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+You can add the new icon in the directory app/images.  It recommended to use png format
+and name the file in lower case with dashes (ex: mysql-logo.png )
+
+## Updating the constants.js file
+The following are the instructions for adding the Constants File:
+1. Add an entry in DIALECTS
+2. Add an entry in SQL_DIALECTS_USING_EDITOR
+3. Add an entry under CONNECTION_CONFIG
+4. Add the location of the logo in
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: 4. Add the location of the logo into LOGOS

File: NEW_CONNECTION.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+and name the file in lower case with dashes (ex: mysql-logo.png )
+
+## Updating the constants.js file
+The following are the instructions for adding the Constants File:
+1. Add an entry in DIALECTS
+2. Add an entry in SQL_DIALECTS_USING_EDITOR
+3. Add an entry under CONNECTION_CONFIG
+4. Add the location of the logo in 
+5. Add the preview query in the PREVIEW_QUERY section
+6. Add an entry in the SAMPLE_DBS
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: 6. Add sample credentials for your connector in the SAMPLE_DBS

File: NEW_CONNECTION.md [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+2. Add an entry in SQL_DIALECTS_USING_EDITOR
+3. Add an entry under CONNECTION_CONFIG
+4. Add the location of the logo in 
+5. Add the preview query in the PREVIEW_QUERY section
+6. Add an entry in the SAMPLE_DBS 
+
+## Updating the Tab.react.js 
+The following are the instructions for adding the Tab React JS file
+1. Add an entry for the connection label
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: Customise the label for your connector here:
```js
        let label;
        if (dialect === DIALECTS.S3) {
            label = `S3 - (${connectionObject.bucket})`;
        } else if (dialect === DIALECTS.APACHE_DRILL) {
            label = `Apache Drill (${connectionObject.host})`;
        } else if (dialect === DIALECTS.APACHE_IMPALA) {
              label = `Apache Impala (${connectionObject.host}:${connectionObject.port})`;
        } else if (dialect === DIALECTS.APACHE_SPARK) {
            label = `Apache Spark (${connectionObject.host}:${connectionObject.port})`;
        } else if (connectionObject.dialect === DIALECTS.ELASTICSEARCH) {
            label = `Elasticsearch (${connectionObject.host})`;
        } else if (connectionObject.dialect === DIALECTS.SQLITE) {
            label = connectionObject.storage;
        } else {
            label = `${connectionObject.database} (${connectionObject.username}@${connectionObject.host})`;
        }
```

File: NEW_CONNECTION.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+## Updating the Settings.react.js
+Update the fetchData function to make the calls to appropriate sessions.js backend.  
+Currently there is Dialect types for SQL, ElasticSearch, S3 and Apache Drill.  If 
+your new data source is an SQL Type you will need to add the new Dialect in the 
+fetchData section for SQL types.  
+
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: I should clean up this file.

For the time being, I'd keep things simple and just ask the developer to update `fetchData` and add their connector dialect to the list in:

```js
        if (contains(connectionObject.dialect, [
                    DIALECTS.APACHE_IMPALA,
                    DIALECTS.APACHE_SPARK,
                    DIALECTS.IBM_DB2,
                    DIALECTS.MYSQL, DIALECTS.MARIADB, DIALECTS.POSTGRES,
                    DIALECTS.REDSHIFT, DIALECTS.MSSQL, DIALECTS.SQLITE,
                    DIALECTS.DATA_WORLD
        ])) {
```

File: NEW_CONNECTION.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * The following function will return a list of schemas.  The return 
+ * array should return a list of array with each array containing the 
+ * following information:
+ * -row[0] schema name
+ * -row[1] column name
+ * -row[2] data type
+ * @param {Object} connection 
+ * @return {Array}
```
Comment (n-riesco) [inline NEW_CONNECTION.md]:
```js
 * @returns {Promise} that resolves to { columnNames: [ 'tablename', 'column_name', 'data_type' ], rows: [[tablename1, columnname1, datatype1], ...]] }
```

File: NEW_CONNECTION.md [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+
+## New Datastore file
+This file should be added in the following location
+backend/persistent/datastores
+
+```javascript
+
+/**
+ * The following function will connect to the datasource
+ * @param {Object} connection
```
Comment (n-riesco) [inline NEW_CONNECTION.md]:
```js
 * @returns {Promise} that resolves when the connection succeeds
```

File: NEW_CONNECTION.md [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+ * @param {Object} connection 
+ * @return {Array}
+ */
+function schemas(connection);
+
+/**
+ * The following function will return a list of tables.  
+ * The return result should an array of strings
+ * @param connection {object}
+ * @return {Array}
```
Comment (n-riesco) [inline NEW_CONNECTION.md]:
```js
 * @returns {Promise} that resolves to an array of table names
```

File: NEW_CONNECTION.md [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+function tables(connection);
+
+/**
+ * The following function will execute a query against the 
+ * data source.  The function should return a tuple which contains 
+ * two elements.  The first is an array of strings which is the column names
+ * The second is a two dimensional array which represents the rows to be displayed
+ * @param {string|object} queryObject
+ * @param {object} connection
+ * @return {object} -- tuple {columnames, rows}
```
Comment (n-riesco) [inline NEW_CONNECTION.md]:
```js
 * @returns {Promise} that resolves to { columnNames, rows }
```

**Comment by n-riesco** [comment]
@shannonlal I've just noticed that you've started to implement the Athena connector.

Shall we change the title of this PR to reflect the change?

**Comment by shannonlal** [comment]
@n-riesco.  Thanks for the comments I can look at these tonight.  With regards to the renaming the pull request.  You can leave this as is.  For the Athena stuff.  I have the backend working and just now have to integrate it into the frontend.  I should have this ready to go in about a week

**Comment by shannonlal** [comment]
@n-riesco.  I noticed that the build is failing due to an issue with DB2 Unit Tests.  I think you had mentioned that this was not related to my work.  Is this till the case?  Is there anything else you need from me for this pull request?  The Athena stuff I will do on a separate pull request.

**Comment by n-riesco** [comment]
@shannonlal For a documentation change, it's OK to ignore CircleCI. But if you want to get rid of the failure, the only thing you need to do is to rebase this PR against master.

**Comment by n-riesco** [comment]
I'll have a look at this PR later today.

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -125,3 +128,14 @@ $ yarn run test-e2e
 - Mac builds are created in TravisCI tests and automatically hosted on [amazon](https://s3.console.aws.amazon.com/s3/buckets/falcon-travis-artifacts/plotly/falcon-sql-client/?region=us-east-1&tab=overview). Select the latest build (largest number *note:* the folders are not necessarily sequential) -> release.zip
 
 Builds are uploaded to https://github.com/plotly/falcon-sql-client/releases.
+
+## Troubleshooting
+The Falcon Configuration information is installed in the users home directory.
```
Comment (n-riesco) [inline CONTRIBUTING.md]: the user's home directory.

File: CONTRIBUTING.md [hunk 19]
```diff
@@ -125,3 +128,14 @@ $ yarn run test-e2e
...
+For example Unix and Mac (~/.plotly/connector).  If you have tried the install
```
Comment (n-riesco) [inline CONTRIBUTING.md]: For Windows `%userprofile%\.plotly\connector\`

File: NEW_CONNECTION.md [hunk 20]
```diff
@@ -0,0 +1,105 @@
+# Add New Data Source
+
+## Images
+It is recommended to add the images as png and a good idea to include.
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: ## Icons

The connector must provide an icon to be displayed on the selection tab.

File: NEW_CONNECTION.md [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+## Images
+It is recommended to add the images as png and a good idea to include.  
+Please verify that the icon looks good with the following CSS specs:
+```css
+ max-width: 100px;
+ max-height: 70px;
+```
+
+## Updating the constants.js file
+The following are the instructions for adding the Constants File:
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: s/adding/updating/

File: NEW_CONNECTION.md [hunk 22]
```diff
@@ -0,0 +1,105 @@
...
+1. Add an entry in DIALECTS
+2. Add an entry in SQL_DIALECTS_USING_EDITOR
+3. Add an entry under CONNECTION_CONFIG
+4. Add the location logo into the LOGOS Constants  
+5. Add the preview query in the PREVIEW_QUERY section
+6. Add sample credentials for your connector in the SAMPLE_DBS 
+
+## Updating the Tab.react.js 
+The following are the instructions for adding the Tab React JS file
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: s/adding/updating/

File: NEW_CONNECTION.md [hunk 23]
```diff
@@ -0,0 +1,105 @@
...
+        } else if (dialect === DIALECTS.NEW_CONNECTION) {
+            //Specify the label
+            label = `NEW CONNECTION (${connectionObject.host})`;
+        }
+    ....
+```
+
+## Updating the Settings.react.js
+Update the fetchData function to make the calls to appropriate sessions.js backend.  
+Add your new connectino to the following list:
```
Comment (n-riesco) [inline NEW_CONNECTION.md]: s/connectino/connection/

**Comment by n-riesco** [comment]
@shannonlal I've made a few small comments after a quick read. I'll merge the PR after you address those.

Again, thank you for writing this up!

**Pull request merged**